Pagination

benkonrath opened the pull request:
This PR is taking over from #38 because it's not possible edit the destination branch. The PR was originally against the version-1.0 branch which has now been merged into master.

dustinfarris commented:
This is just waiting on docs, right?

benkonrath commented:
Yeah. I'm hoping to have time for this later in the week.

benkonrath commented:
Ok, I found some time to do a first version of the docs. I've asked for contributions for docs about integration with ember-cli-pagination since I don't have time to figure it out right now.

Feel free to give the docs a thorough edit for style, tone, bad grammar or whatever. I won't be offended. :)

benkonrath commented:
I need to add an entry in the Changelog. I'll do that right now.

dustinfarris commented on docs/pagination.md:
````diff
@@ -0,0 +1,71 @@
+
+```js
+if (meta.next) {
+  store.find('post', {page: meta.next}
````
Missing a closing paren here.

benkonrath replied:
Thanks, I just fixed it.

dustinfarris commented on CHANGELOG.md:
```diff
@@ -1,6 +1,13 @@
 ember-django-adapter Changelog
 ==============================
 
+Pending
```
Let's call this section "Master", mimicking ember-cli's format.

dustinfarris commented on docs/pagination.md:
```diff
@@ -0,0 +1,71 @@
+
+Pagination is supported using the metadata support that is built into Ember Data. The
+metadata from Django REST Framework paginated list views is updated on every request
+to sever.
```
"to the server." ?

benkonrath replied:
Yep. It's sort of funny. What exactly am I severing? :)

dustinfarris commented on docs/pagination.md:
```diff
@@ -0,0 +1,71 @@
+
+## Retrieving the Metadata
+
+To get a page of records simply run a find request with the `page` query param:
```
"page of records, simply"

dustinfarris commented on docs/pagination.md:
````diff
@@ -0,0 +1,71 @@
+});
+```
+
+It's possible to override the server-side page size by also including the `page_size`
````
Has this been tested on DRF yet?  Are we sure `page_size` works?

benkonrath replied:
http://www.django-rest-framework.org/api-guide/pagination/

I just tested this and it seems to be off by default. I'll adjust the docs on this point.

benkonrath replied:
Given that page_size is off by default in DRF, I think we should add a pageSize property to the metadata. It's difficult for applications to get the page size but it's easy to do in the serializer. The page size is needed to calculate the number of pages. What do you think?

dustinfarris replied:
Users may just set the `PAGINATE_BY` setting in DRF and forego the `PAGINATE_BY_PARAM` setting altogether—which is fine, and simpler.  So this should be optional and off by default on our end as well.  

Can we make this customizable, so that the user can specify the param name, if any, just like they can in DRF?

benkonrath replied:
It's already possible to use any param name for page size. ED just passes the param to the backend and EDA doesn't change that. (I'll document this as well)

The issue is that if PAGINATE_BY_PARAM is disabled (which is the default), the application doesn't know the page size. Of course front-end ember applications can always hardcode the PAGINATE_BY value but that is just a copy of the back-end DRF setting for the view. It seems more flexible if the serializer sets the page size as metadata so applications can "read" the page size. If the back-end DRF setting changes, the correct page size will be automatically in the metadata. This isn't an issue if an application uses 'page_size' explicitly because the front-end application is in charge of the setting in this case, not the back-end.

benkonrath replied:
It turns out there are no pagination related defaults in DRF:
https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/settings.py#L61

I usually set `paginated_by` on the view to enable pagination and I usually use different page sizes across my paginated views  (i.e. I don't use the `PAGINATE_BY` setting). This is one reason why I'm pushing for setting the `pageSize` in the metadata.

I'm going to add the `pageSize` to the metadata for you to review. I'm pushing hard for this point because I think it makes things easier for application developers. Do you think having the `pageSize` in the metadata is a bad idea?

dustinfarris commented on docs/pagination.md:
```diff
@@ -0,0 +1,71 @@
+
+## Metadata Properties
+
+The metadata consists of three properties which gives applications enough information
```
"three properties that give the applications"

dustinfarris commented:
Documentation looks good.  I left a couple nitpicking comments.

benkonrath commented:
I've updated the documentation. I added all of your corrections and added a new section about the page size that I need to finish once we have thing solidified.

benkonrath commented:
The two things remaining:

* Decide if having the `pageSize` in the metadata is a good idea.
* Finish documentation about page size

benkonrath commented:
After some thought, I think including the `pageSize` in the metadata isn't a good idea because of the issue with not having an accurate value on the last page. We can always add it later if there is a need but I think the metadata is useful for pagination without it. I'm going to remove the last commit and update the docs about the `page_size` DRF query param.

benkonrath commented:
Ok, I've re-worked the docs for page_size. @dustinfarris Can you look at the pagination docs one more time?

dustinfarris commented:
@benkonrath pinged you in IRC

dustinfarris commented:
:+1:
